### DISCUSSION TITLE
记一次奇怪的编译错误

### DISCUSSION THREAD

**Discussion by Manicsteiner** [MaaAssistantArknights/MaaAssistantArknights#10876] [General]
不是bug反馈，大概
省流：不要装多个不同版本的 MSVC v143 生成工具

自 https://github.com/MaaAssistantArknights/MaaAssistantArknights/commit/e5874b581bad59e6bca68e5d242ee1046dae0bef 更新了MaaDeps之后，一台电脑上MAA编译始终报错，另一台一切正常。
错误大概是说 MaaCore 要用到的 cpr.lib 有一条引用函数找不到。

> 严重性	代码	说明	项目	文件	行	禁止显示状态	详细信息
> 错误	LNK2001	无法解析的外部符号 _Cnd_timedwait_for	MaaCore	C:\Storage\Developer\MAA\MaaAssistantArknights\src\MaaCore\cpr.lib(threadpool.cpp.obj)	1		
> 错误	LNK1120	1 个无法解析的外部命令	MaaCore	C:\Storage\Developer\MAA\MaaAssistantArknights\x64\Release\MaaCore.dll	1

> 严重性	代码	说明	项目	文件	行	禁止显示状态	详细信息
> 错误	LNK2019	无法解析的外部符号 _Cnd_timedwait_for，函数 `"public: enum std::cv_status __cdecl std::condition_variable::wait_until<struct std::chrono::steady_clock,class std::chrono::duration<__int64,struct std::ratio<1,1000000000> > >(class std::unique_lock<class std::mutex> &,class std::chrono::time_point<struct std::chrono::steady_clock,class std::chrono::duration<__int64,struct std::ratio<1,1000000000> > > const &)" (??$wait_until@Usteady_clock@chrono@std@@V?$duration@_JU?$ratio@$00$0DLJKMKAA@@std@@@23@@condition_variable@std@@QEAA?AW4cv_status@1@AEAV?$unique_lock@Vmutex@std@@@1@AEBV?$time_point@Usteady_clock@chrono@std@@V?$duration@_JU?$ratio@$00$0DLJKMKAA@@std@@@23@@chrono@1@@Z) 中引用了该符号	MaaCore	C:\Storage\Developer\MAA\MaaAssistantArknights\src\MaaCore\cpr.lib(threadpool.cpp.obj)	`1		
> 错误	LNK1120	1 个无法解析的外部命令	MaaCore	C:\Storage\Developer\MAA\MaaAssistantArknights\x64\Debug\MaaCore.dll	1

于是就各种重下依赖、重新手动解压，各种删编译缓存，都不能解决问题。
于是我决定整个烂活，从正常编译的那台电脑上复制所有相关文件过来，看看能不能编译。然后失败了。
不过这也是个好消息，这说明是电脑的环境问题，和MAA无关。同时也是个坏消息，因为这台电脑安装了更多的VS组件，也有更多的其他工具在系统变量里。

于是先对比了一下是不是少了什么VS组件，没有发现，然后在多出来的组件里看看，最可疑也最没用的就是 MSVC v143 - VS2022 C++生成工具 v14.38-17.8 的旧版本了。于是把它卸载了，再编译，就一切正常了。

总之是个很奇怪的问题，毕竟这个电脑上还装了 v142 和 v141 的生成工具，似乎不会互相影响，但 v143 14.38 就会和 v143最新 14.40 产生影响，但是编译其他C++程序时就没有见过这种问题，更新 MaaDeps 之前也正常共存了几个月。

总之总之这个问题就这样解决了。

**Comment by chisato2233** [comment]
我了个豆，完全一样的问题，还好有github的issue啊，看了一眼自己的工具集果然有msvc v143 的14.38-17.8，这果然是bug吧。。。。。。

**Comment by Tokisaki-Galaxy** [comment]
还好我用的是action编译，本地配环境好多坑啊（
